### PR TITLE
PPL: Execute Explore PPL logs format queries through the backend

### LINKED ISSUE #200: Backend: Migrate PPL Logs queries to the backend
Prerequisite: https://github.com/grafana/opensearch-datasource/issues/226

PPL Logs are processed into dataFrames on the FE, and as timeSeries on the BE. They should probably be separated out.
ES plugin doesn't support PPL queries, but maybe the [Logs PR](https://github.com/grafana/grafana/pull/63647) for Lucene could be something to look at for reference.

Some general guidelines are:

* Create visualization to compare between frontend and backend flows, preferably in a shared Cloud dashboard e.g. clouddatasources.grafana.net 
* Unit tests in the frontend should be copied or updated in the backend to assert the same behavior and add unit tests when there are none.
    * Check Elasticsearch repo for related tests – helpful when there are no OpenSearch tests in the frontend or backend
* We should match OpenSearch's backend behavior to the current frontend behavior, but be aware it may not always make sense. Elasticsearch's current behavior is also another resource for how OpenSearch should behave.
    * Let's be pragmatic about this – our foremost goal is the migration of OpenSearch's frontend behavior. We can aim for some Elasticsearch feature parity, but only _within reason_.
* Often there is a PR in the Elasticsearch backend data source migration [https://github.com/grafana/grafana/issues/54011](https://github.com/grafana/grafana/issues/54011) which is 75% of what OpenSearch will need
* Migrate or implement: i.e. make any changes to code in backend
* Use the dashboard created above to compare visualization results (data frame comparison)
    * Can also inspect data frame on a more detailed basis
   
  
Note: ([Similarly](https://github.com/grafana/opensearch-datasource/issues/199) to Lucene Logs of the FE) It looks like PPL Logs are not completely correctly processed on the Frontend, as the int field is passed as string. This isn't the case on the BE, as seen from the screenshot. 
Screenshot 2023-06-12 at 1 31 26 PM
Screenshot 2023-06-12 at 1 32 18 PM

### PR DESCRIPTION
**Pull request by kevinwcyu**
**What this PR does / why we need it**:

This executes PPL queries with the `format` set to `logs` from Explore to be executed on the backend.

To test:
- Go to Explore
- Set "Query" to `PPL`
- Set "Format" to `Logs`
- Use the "Logs" panel
- Enter your PPL query: you can find some example queries [here](https://clouddatasources.grafana.net/goto/vMOq6OmIg?orgId=1) under the "PPL Logs" row

**PPL logs queries will only run in the backend in the Explore view** in order to limit this migration's impact to dashboards. Please [let us know](https://github.com/grafana/opensearch-datasource/issues/new?assignees=&labels=datasource%2FOpenSearch%2Ctype%2Fbug&projects=&template=bug_report.md) if you observe some unwanted behaviour while executing PPL Log queries in Explore.

**Which issue(s) this PR fixes**:

Fixes #200 

**Special notes for your reviewer**:

**There are some differences between the frontend execution and backend execution**:
- Fields are sorted alphabetically. The only exceptions are the time field and the log message field. The time field is the first field, and the log message field is the second field if it is configured in the data source settings.
- Fields are no longer treated as strings, instead they are typed properly. E.g. the frontend would treat `false` as a string, but the backend will now treat is as a bool.

**Testing ad-hoc filters**:
The logs panel doesn't provide a button to apply ad-hoc filters. Instead you'll need to test this from a dashboard with the following procedure:
- Enable PPL log queries to be sent to the backend by removing `request.app === CoreApp.Explore` from line 546 in `src/datasource.ts`
- Create a new visualization
- Set query to "PPL" and format to "Logs"
- Use the "Table" panel
- Run a query (you can just leave the query field blank if you have an index configured)
- Wait until some data is returned and then hover over one of the cells in the table panel and click the "+" icon.
- You can change to the "Logs" panel
- Alternatively, you can just add an ad-hoc filter template variable.

### CODE REVIEW

File: pkg/opensearch/ppl_response_parser.go [hunk 1]
```diff
@@ -53,7 +55,69 @@ func (rp *pplResponseParser) parseTimeSeries() (*backend.DataResponse, error) {
...
+	frame := data.NewFrame("", fields...)
+	if frame.Meta == nil {
+		frame.Meta = &data.FrameMeta{}
+	}
+	frame.Meta.PreferredVisualization = data.VisTypeLogs
+
+	queryRes.Frames = append(queryRes.Frames, frame)
+
+	return queryRes, nil
+}
```
Comment (fridgepoet) [inline pkg/opensearch/ppl_response_parser.go:117]: I wonder to what extent we should try to mutualize some of this code with the Lucene code, what do you think? (maybe in a following PR)

I'd like to avoid a situation where we have to make changes on "both sides." At the same time I don't want to introduce code that is too rigid.

**Comment by fridgepoet** [comment]
> Fields are sorted alphabetically. The only exceptions are the time field and the log message field. The time field is the first field, and the log message field is the second field if it is configured in the data source settings.

What's the reason for this?

---
What do you think of not restricting this to Explore? Too risky?

**Review by fridgepoet — APPROVED** [review]

File: pkg/opensearch/client/ppl_request.go [hunk 2]
```diff
@@ -31,11 +31,17 @@ func (b *PPLRequestBuilder) AddPPLQueryString(timeField, to, from, querystring s
...
 
+	trimmedQuerystring := strings.TrimSpace(querystring)
 	// Sets a default query if the query string is empty
-	if len(strings.TrimSpace(querystring)) == 0 {
+	if len(trimmedQuerystring) == 0 {
 		querystring = fmt.Sprintf("source = %s", b.index)
 	}
 
+	// Set a source index if the query string is not empty, but includes ad-hoc filters.
+	if !strings.HasPrefix(trimmedQuerystring, "source") && !strings.HasPrefix(trimmedQuerystring, "search") {
```
[the remark this answers is not in the document]
Comment (kevinwcyu) [reply pkg/opensearch/client/ppl_request.go:41]: no real reason. both conditions appeared to be the same to me. i've updated it though to check for "| where `", since that matches what the the frontend does when adding ad-hoc filters

File: pkg/opensearch/ppl_handler.go [hunk 3]
```diff
@@ -41,8 +44,10 @@ func (h *pplHandler) executeQueries() (*backend.QueryDataResponse, error) {
 		if err != nil {
 			return nil, err
 		}
-		rp := newPPLResponseParser(res)
-		queryRes, err := rp.parseTimeSeries()
+
+		query := h.queries[refID]
+		rp := newPPLResponseParser(res, query)
+		queryRes, err := rp.parseResponse(h.client.GetConfiguredFields())
```
[the remark this answers is not in the document]
Comment (kevinwcyu) [reply pkg/opensearch/ppl_handler.go:50]: sounds good.

File: pkg/opensearch/ppl_response_parser.go (hunk 1 above)
Comment (kevinwcyu) [reply pkg/opensearch/ppl_response_parser.go:117]: this particular part of the code that processes logs is relatively short so i don't think there's too much of a problem with having to make changes to both sides.

we can look at sharing some of the code with the lucene logs side though. i think after finishing the rest of the query migrations would be a good time to do so, maybe when we are actually working on any changes to the logs processing.

**Comment by kevinwcyu** [comment]
> > Fields are sorted alphabetically. The only exceptions are the time field and the log message field. The time field is the first field, and the log message field is the second field if it is configured in the data source settings.
> 
> What's the reason for this?

for sorting alphabetically or why the time field and log message fields are placed in the first and second field?

if it's for sorting alphabetically: i saw in the opensearch dashboard that the fields are sorted alphabetically and i think it's easier to look for certain fields when looking through logs. e.g. for the current implementation https://clouddatasources.grafana.net/d/0Km_Gnf4k2/opensearch-non-regression?orgId=1&editPanel=64 if you expand the details, the fields are not sorted so it's a bit harder to find the field you're looking for at a quick glance if you're just exploring the data.

if it's about the time field and log message field positions, this is the logic that's currently present in the frontend. for the "log message field", if it's specified in the config, it should be the second field so that the logs panel, knows to treat that as the default message. the first time field is also treated as the timestamp, so that's why it's set as the first field.

> 
> What do you think of not restricting this to Explore? Too risky?

yea, think it'd be good to let this sit in explore for a bit first.
